driver-core/src/main/com/mongodb/internal/connection/IamAuthenticator.java

```diff
@@ -0,0 +1,203 @@
+            this.randomStringGenerator = randomStringGenerator;
+        }
+
+        @Override
+        public String getMechanismName() {
+            AuthenticationMechanism authMechanism = credential.getAuthenticationMechanism();
+            if (authMechanism == null) {
+                throw new IllegalArgumentException("Authentication mechanism cannot be null");
+            }
+            return authMechanism.getMechanismName();
+        }
+
+        @Override
+        public boolean hasInitialResponse() {
+            return true;
+        }
+
+        @Override
+        public byte[] evaluateChallenge(final byte[] challenge) throws SaslException {
+            step++;
+            if (step == 0) {
+                return computeClientFirstMessage();
+            }
+            if (step == 1) {
+                return computeClientFinalMessage(challenge);
+            } else {
+                throw new SaslException(format("Too many steps involved in the %s negotiation.", getMechanismName()));
+            }
+        }
+
+        @Override
+        public boolean isComplete() {
+            return step == 1;
+        }
+
+        @Override
+        public byte[] unwrap(final byte[] bytes, final int i, final int i1) throws SaslException {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public byte[] wrap(final byte[] bytes, final int i, final int i1) throws SaslException {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public Object getNegotiatedProperty(final String s) {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public void dispose() throws SaslException {
+            // nothing to do
+        }
+
+        private byte[] decodeUTF8(final String str) throws SaslException {
+            try {
+                return str.getBytes("UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                throw new SaslException("UTF-8 is not a supported encoding.", e);
+            }
+        }
+
+        private byte[] computeClientFirstMessage() throws SaslException {
+            clientNonce = randomStringGenerator.generate(RANDOM_LENGTH);
+            BsonDocument document = new BsonDocument()
+                    .append("r", new BsonBinary(decodeUTF8(clientNonce)))
+                    .append("p", new BsonInt32('n'));
+            return toBson(document);
+        }
+
+        private byte[] computeClientFinalMessage(final byte[] serverFirst) throws SaslException {
+            final BsonDocument document = new RawBsonDocument(serverFirst);
+            final String host = document.getString("h").getValue();
+
+            final byte[] serverNonce = document.getBinary("s").getData();
+            if (serverNonce.length != 64 || !Arrays.equals(Arrays.copyOf(serverNonce, 32), decodeUTF8(clientNonce))) {
```

Comment:
Instead of `32` and `64`, I would use `RANDOM_LENGTH` and create a new constant respectively.

Reply:
Done. I just used `RANDOM_LENGTH * 2` for 64 for clarity.